default/python/AI/ProductionAI.py

```diff
@@ -2033,3 +1697,285 @@ def _build_military_command() -> float:
         # default selection should prefer the capital, unless its species dislikes it.
         return _try_enqueue(building_type, get_inhabited_planets())
     return 0.0
+
+
+def _build_basic_shipyards() -> Tuple[
+    List[PlanetId], Dict[PlanetId, SystemId], Dict[SystemId, List[Tuple[PlanetId, float]]]
+]:
+    """
+    Consider building basic ship yards and also determine some value needed for other shipyard buildings.
+    Returns queued_shipyard_pids, colony_systems, top_pilot_systems.
+    """
+    building_type = BuildingType.SHIPYARD_BASE
+    universe = fo.getUniverse()
+    queued_shipyard_pids = building_type.queued_at()
+    system_colonies = {}
+    colony_systems = {}
+    empire_species = get_empire_planets_by_species()
+    for spec_name in get_colony_builders():
+        if not get_colony_builder_locations(spec_name) and (
+            spec_name in empire_species
+        ):  # not enough current shipyards for this species #TODO: also allow orbital incubators and/or asteroid ships
+            for pid in get_empire_planets_with_species(
+                spec_name
+            ):  # SP_EXOBOT may not actually have a colony yet but be in empireColonizers
+                if pid in queued_shipyard_pids:
+                    break  # won't try building more than one shipyard at once, per colonizer
+            else:
+                # no queued shipyards: get planets with target pop >=3 and
+                # queue a shipyard on the one with the biggest current population
+                planets = (universe.getPlanet(x) for x in get_empire_planets_with_species(spec_name))
+                pops = sorted(
+                    (planet_.initialMeterValue(fo.meterType.population), planet_.id)
+                    for planet_ in planets
+                    if (planet_ and planet_.initialMeterValue(fo.meterType.targetPopulation) >= 3.0)
+                )
+                pids = [pid for pop, pid in pops if building_type.can_be_produced(pid)]
```

Comment:
You could extract conditions to the function, and use `key` for sorting.

```suggestion
                def pids_filter(planet_):
                    return (
                            planet_ and
                            planet_.initialMeterValue(fo.meterType.targetPopulation) >= 3.0 and
                            building_type.can_be_produced(planet_.id
                          )
    
                    def sort_key(planet_):
                    return planet_.initialMeterValue(fo.meterType.population)
    
                pids = sorted(
                    (planet_.id for planet_ in planets if pids_filter(planet_)), key=sort_key)
```